### PR TITLE
Lecturer availability

### LINKED ISSUE #37: Refactor lecturer/staff table to 3NF with course and availability junction tables
Parent issue: #9 Set availability

Description:
## Context
The current `staff` table stores courses as JSON arrays, violating 1NF. Based on ADR-004's decision to normalize student-course data, apply the same pattern to lecturer data.

## What changes
- Remove `courses` JSON column from `staff` table
- Create `staff_courses(staff_number, course_code)` junction table
- Create `lecturer_availability(staff_number, day_of_week, start_time, end_time)` table
- Add foreign key constraints referencing `staff` and `courses`

## Acceptance Tests
- [x] `staff` table has no JSON columns
- [x] `staff_courses` junction table exists with correct FK constraints
- [ ] `lecturer_availability` table exists with day-of-week constraint (Mon-Fri)
- [ ] Migration rolls back cleanly
- [ ] Existing data migrates without loss

Partially superseded by the schema work landing in #43.

The `courses` JSON column removal from `staff` and the `staff_courses` 
junction table are being handled as part of the #43 branch 
(departments + seed restructure), since both changes touch `createSchema.sql` 
and doing them in separate PRs would cause merge conflicts.

What remains exclusively in this PR:
- `lecturer_availability(staff_number, day_of_week, start_time, end_time)` 
  table — this is NOT in the #43 branch and still needs to land here.
- ADR-XXX documenting the lecturer availability schema decision.

Acceptance tests 1 and 2 (staff has no JSON columns, staff_courses exists) 
will be green after #43 merges. Tests 3–5 (lecturer_availability table, 
rollback, data migration) are still this PR's responsibility.

Suggest rebasing this branch on main after #43 merges before opening the PR.

## ADR
Requires ADR-XXX: Lecturer and availability schema normalization

### PR DESCRIPTION
**Pull request by Sarcastic-Sunflower**
closes #37 

# Lecturer availability Description

> This PR introduces the `lecturer_availability` feature, including schema updates, seed data, documentation, and test coverage.

## Database Changes
Created `lecturer_availability` table to store lecturer time-slot availability

### Enforced constraints for:
- Valid weekday values (Mon–Fri)
- Proper time formatting (HH:MM)
- Logical time ordering (end_time > start_time)
- Positive booking duration limits
- Maximum student capacity per slot
- Minimum venue name length
- Added foreign key relationship to staff table for referential integrity

### Seed Data
Added lightweight seed data for:
- Initial lecturer availability records
- Ensured seed data supports realistic testing scenarios without overloading the database

### Documentation
Added a new ADR documenting:
- Design decisions for lecturer availability structure
 - Constraint choices and rationale

### Testing
Implemented unit tests for `lecturer_availability`
Covered key constraint validations:
- Valid insertions
- Invalid day-of-week values
- Time format validation
- Logical time constraints
- Capacity and booking limits
- Foreign key enforcement
- Primary key uniqueness

### Notes
Database schema and seed structure are now aligned with test expectations
Foreign key enforcement is enabled for the in-memory test environment

### CODE REVIEW

File: database/createSchema.sql [hunk 1]
```diff
@@ -94,3 +95,24 @@ CREATE TABLE consultation_attendees (
   FOREIGN KEY (const_id)       REFERENCES consultations(const_id) ON DELETE CASCADE,
   FOREIGN KEY (student_number) REFERENCES students(student_number) ON DELETE CASCADE
 );
+
+CREATE TABLE lecturer_availablity (
+  availability_id TEXT PRIMARY KEY,
+  staff_number TEXT NOT NULL,
+  day_of_week TEXT NOT NULL CHECK (day_of_week IN ('Mon','Tue','Wed','Thu','Fri')),
```
Comment (Aditya-Raghunandan) [inline database/createSchema.sql:102]: `lecturer_availablity` — the table name has a typo: "availablity" instead of "availability".
This will propagate into every query, seed insert, and test. Fix the name before merging or
document it as a known defect you'll track in a follow-up issue.

File: database/createSchema.sql [hunk 2]
```diff
@@ -94,3 +95,24 @@ CREATE TABLE consultation_attendees (
...
+    CHECK (max_booking_min > 0 AND max_booking_min <= 480),
+
+  max_number_of_students INTEGER NOT NULL DEFAULT 1 
+    CHECK (max_number_of_students > 0),
+
+  venue TEXT NOT NULL CHECK(length(venue) >= 3),
+
+  CHECK (end_time > start_time),
+
+  FOREIGN KEY (staff_number) REFERENCES staff(staff_number) ON DELETE CASCADE
```
Comment (Aditya-Raghunandan) [inline database/createSchema.sql:117]: good use of foreign keys

File: database/seedVitalInfo.sql [hunk 3]
```diff
@@ -296,4 +296,10 @@ INSERT OR IGNORE INTO staff_courses (staff_number, course_code) VALUES
...
 INSERT OR IGNORE INTO students (student_number, name, email, password, degree_code) VALUES
-  (1234567, 'Aditya Raghunandan', 'aditya@students.wits.ac.za', 'pass', 'BSCENGINFO');
\ No newline at end of file
+  (1234567, 'Aditya Raghunandan', 'aditya@students.wits.ac.za', 'pass', 'BSCENGINFO');
+
+INSERT OR IGNORE INTO lecturer_availablity (availability_id, staff_number, day_of_week, start_time, end_time, max_booking_min, max_number_of_students, venue) VALUES
+  ('A000356-1', 'A000356', 'Mon', '10:00', '11:00', 60, 5, 'Room 101'),
+  ('A000357-1', 'A000357', 'Tue', '11:00', '13:00', 120, 10, 'Room 102'),
+  ('A000356-2', 'A000356', 'Wed', '10:00', '14:00', 180, 11, 'Room 103'),
+  ('A000357-2', 'A000357', 'Thu', '11:00', '14:00', 120, 6, 'Room 104');
```
Comment (Aditya-Raghunandan) [inline database/seedVitalInfo.sql]: The staff_number + increment pattern for availability_id is readable but requires the 
application to manage sequencing. Consider whether a UUID or autoincrement integer PK 
would be safer under concurrent requests. Fine for now given our small scale.

File: database/seedVitalInfo.sql [hunk 4]
```diff
@@ -296,4 +296,10 @@ INSERT OR IGNORE INTO staff_courses (staff_number, course_code) VALUES
 
 -- ─── Test student ─────────────────────────────────────────────────────────────
 INSERT OR IGNORE INTO students (student_number, name, email, password, degree_code) VALUES
-  (1234567, 'Aditya Raghunandan', 'aditya@students.wits.ac.za', 'pass', 'BSCENGINFO');
\ No newline at end of file
+  (1234567, 'Aditya Raghunandan', 'aditya@students.wits.ac.za', 'pass', 'BSCENGINFO');
+
+INSERT OR IGNORE INTO lecturer_availablity (availability_id, staff_number, day_of_week, start_time, end_time, max_booking_min, max_number_of_students, venue) VALUES
```
Comment (Aditya-Raghunandan) [inline database/seedVitalInfo.sql]: The seed inserts use the misspelled table name. If the schema typo is fixed, the seed will
break. Fix both together in the same commit.

File: documentation/ADR's/ADR-008-lecturerAvailability-schema [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+|---|---|
+| `availability_id` | PK to table - Consists of staff_number and  an incremental number|
+| `staff_number` | FK to the lecturer who owns this window |
+| `day_of_week` | Recurring weekday - Mon–Fri only, enforced by constraint |
+| `start_time` / `end_time` | Window bounds in HH:MM, consistent with `consultations.consultation_time` |
+| `max_booking_min` | Maximum duration a student may request when booking within this window |
+
+This table stores **one fact**: when a lecturer is generally available and
+for how long students may book. It contains no date, no venue, no student
+reference — those belong on `consultations` when an actual booking is made.
```
Comment (Aditya-Raghunandan) [inline documentation/ADR's/ADR-008-lecturerAvailability-schema:60]: here

File: documentation/ADR's/ADR-008-lecturerAvailability-schema [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  start_time TEXT NOT NULL CHECK (start_time GLOB '[0-9][0-9]:[0-9][0-9]'),
+  end_time TEXT NOT NULL CHECK (end_time GLOB '[0-9][0-9]:[0-9][0-9]'),
+  
+  max_booking_min INTEGER NOT NULL DEFAULT 60 CHECK (max_booking_min > 0 AND max_booking_min <= 480),
+  CHECK (end_time > start_time)
+
+  max_number_of_students   INTEGER NOT NULL DEFAULT 1 CHECK (max_number_of_students > 0),
+  venue                    TEXT NOT NULL CHECK(length(venue) >= 3),
+
+  FOREIGN KEY (lecturer_id) REFERENCES staff(staff_number) ON DELETE CASCADE,
```
Comment (Aditya-Raghunandan) [inline documentation/ADR's/ADR-008-lecturerAvailability-schema]: The ADR prose and reasoning are solid. Two fixes needed:
1. The SQL block in the Decision section has `FOREIGN KEY (lecturer_id)` and a trailing 
   comma before `);` — neither matches the actual implementation. Update to match the real DDL.
2. The column table omits `max_number_of_students` and `venue`. The Consequences section 
   also states the table contains "no venue" which contradicts the schema.

File: documentation/ADR's/ADR-008-lecturerAvailability-schema [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  day_of_week TEXT NOT NULL CHECK (day_of_week IN ('Mon','Tue','Wed','Thu','Fri')),
+  
+  start_time TEXT NOT NULL CHECK (start_time GLOB '[0-9][0-9]:[0-9][0-9]'),
+  end_time TEXT NOT NULL CHECK (end_time GLOB '[0-9][0-9]:[0-9][0-9]'),
+  
+  max_booking_min INTEGER NOT NULL DEFAULT 60 CHECK (max_booking_min > 0 AND max_booking_min <= 480),
+  CHECK (end_time > start_time)
+
+  max_number_of_students   INTEGER NOT NULL DEFAULT 1 CHECK (max_number_of_students > 0),
+  venue                    TEXT NOT NULL CHECK(length(venue) >= 3),
```
Comment (Aditya-Raghunandan) [inline documentation/ADR's/ADR-008-lecturerAvailability-schema:47]: we do have venue and max_number of students

File: tests/unit/availability-db.test.js [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+  afterAll(() => {
+    db.close()
+  })
+
+  const insertAvailability = (values) => {
+    return db.prepare(`
+      INSERT INTO lecturer_availablity 
+      VALUES (?, ?, ?, ?, ?, ?, ?, ?)
+    `).run(...values)
```
Comment (Aditya-Raghunandan) [inline tests/unit/availability-db.test.js:36]: The helper uses positional `?` parameters with no column names. If a column is added or 
reordered in the schema, tests will silently insert wrong data without failing. Specify 
columns explicitly:

  INSERT INTO lecturer_availablity 
    (availability_id, staff_number, day_of_week, start_time, end_time,
     max_booking_min, max_number_of_students, venue)
  VALUES (?, ?, ?, ?, ?, ?, ?, ?)

File: tests/unit/availability-db.test.js [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        'Mon',
+        '10:00',
+        '11:00',
+        60,
+        0,
+        'Room 105'
+      ])
+    }).toThrow()
+  })
+
```
Comment (Aditya-Raghunandan) [inline tests/unit/availability-db.test.js:173]: these two tests above , tests 0 (reject) and 500 (reject) but never tests 480, which the CHECK allows. 
Add a test confirming 480 does NOT throw, to verify the upper bound is inclusive.

File: tests/unit/availability-db.test.js [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+        'A000356',
+        'Mon',
+        '10:00',
+        '11:00',
+        60,
+        5,
+        'R1'
+      ])
+    }).toThrow()
+  })
```
Comment (Aditya-Raghunandan) [inline tests/unit/availability-db.test.js:187]: 'R1' (length 2) is correctly rejected. But there is no test confirming that a 3-character 
venue like 'R10' is accepted. Add a positive boundary test for the minimum valid venue length.

**Comment by Aditya-Raghunandan** [comment]
Good first pass at this feature — the test suite covers all the right constraint categories 
and the ADR reasoning is well argued. Three things to address:

Blocking:
- Table name typo `lecturer_availablity` throughout schema, seed, and tests — needs to be 
  fixed to `lecturer_availability` consistently before merging.

Non-blocking (can be follow-up issues):
- INSERT helper should name columns explicitly to avoid silent breakage on schema changes.
- ADR Decision section: the SQL block has `FOREIGN KEY (lecturer_id)` and a trailing comma 
  — neither matches the actual implementation. Also the sentence "It contains no date, no 
  venue" contradicts the schema which does include `venue`. Small fixes.
- ADR column table omits `max_number_of_students` and `venue`.
- No boundary tests for `max_booking_min = 480` (valid) or `venue = 'R10'` (valid, 3 chars).

Fix the typo and push — happy to merge after that.

**Review by Aditya-Raghunandan — CHANGES_REQUESTED** [review]
Good first pass at this feature — the test suite covers all the right constraint categories 
and the ADR reasoning is well argued. Three things to address:

Blocking:
- Table name typo `lecturer_availablity` throughout schema, seed, and tests — needs to be 
  fixed to `lecturer_availability` consistently before merging.

Non-blocking (can be follow-up issues):
- INSERT helper should name columns explicitly to avoid silent breakage on schema changes.
- ADR Decision section: the SQL block has `FOREIGN KEY (lecturer_id)` and a trailing comma 
  — neither matches the actual implementation. Also the sentence "It contains no date, no 
  venue" contradicts the schema which does include `venue`. Small fixes.
- ADR column table omits `max_number_of_students` and `venue`.
- No boundary tests for `max_booking_min = 480` (valid) or `venue = 'R10'` (valid, 3 chars).

Fix the typo and push — happy to merge after that.
